File: Text/Encoders/EncodingSingleByte.cs
using System;
using System.IO;
using PeterO;

using PeterO.Text;

namespace PeterO.Text.Encoders {
 internal class EncodingSingleByte : ICharacterEncoding {
   private class Decoder : ICharacterDecoder {
      private readonly int[] codepoints;

      public Decoder(int[] codepoints) {
        this.codepoints = codepoints;
      }

     public int ReadChar(IByteReader transform) {
       int b = transform.ReadByte();
       return (b < 0) ? (-1) : ((b < 0x80) ? b : this.codepoints[b - 0x80]);
    }
  }

   private class Encoder : ICharacterEncoder {
      private readonly int[] codepoints;

      public Encoder(int[] codepoints) {
        this.codepoints = codepoints;
      }

    public int Encode(
     int c,
     IWriter output) {
       if (c < 0) {
 return -1;
}
         if (c < 0x80) {
           output.WriteByte((byte)c);
           return 1;
         }
         for (var i = 0; i < this.codepoints.Length; ++i) {
           if (this.codepoints[i] == c) {
             output.WriteByte((byte)(i + 0x80));
             return 1;
           }
         }
         return -2;
    }
  }

    private readonly Encoder encoder;
    private readonly Decoder decoder;

    public EncodingSingleByte(int[] codepoints) {
        if (codepoints == null) {
  throw new ArgumentNullException("codepoints");
}
        if (codepoints.Length != 128) {
  throw new ArgumentException("codepoints.Length (" + codepoints.Length +
    ") is not equal to " + 128);
}
this.encoder = new Encoder(codepoints);
this.decoder = new Decoder(codepoints);
      }

  public ICharacterDecoder GetDecoder() {
    return this.decoder;
  }

  public ICharacterEncoder GetEncoder() {
    return this.encoder;
  }
 }
}

File: src/main/java/com/upokecenter/text/encoders/EncodingSingleByte.java
package com.upokecenter.text.encoders;

import java.io.*;
import com.upokecenter.util.*;

import com.upokecenter.text.*;

 public class EncodingSingleByte implements ICharacterEncoding {
   private static class Decoder implements ICharacterDecoder {
      private int[] mapping;

      public Decoder (int[] mapping) {
        this.mapping = mapping;
      }

     public int ReadChar(ITransform transform) {
       int b = transform.read();
       return (b < 0) ? (-1) : ((b < 0x80) ? b : this.mapping.get(b - 0x80));
    }
  }

   private static class Encoder implements ICharacterEncoder {
      private int[] mapping;

      public Encoder (int[] mapping) {
        this.mapping = mapping;
      }

    public int Encode(
     int c,
     InputStream output) {
       if (c < 0) {
 return -1;
}
         if (c < 0x80) {
           output.write((byte)c);
           return 1;
         }
         for (int i = 0; i < this.mapping.length; ++i) {
           if (this.mapping.get(i) == c) {
             output.write((byte)(i + 0x80));
             return 1;
           }
         }
         return -2;
    }
  }

   private Encoder encoder;
   private Decoder decoder;

  public EncodingSingleByte (int[] mapping) {
        if (mapping == null) {
  throw new NullPointerException("mapping");
}
        if (mapping.length != 128) {
  throw new IllegalArgumentException("mapping.length (" + mapping.length +
    ") is not equal to " + 128);
}
this.encoder = new Encoder(mapping);
this.decoder = new Decoder(mapping);
      }

  public ICharacterDecoder GetDecoder() {
    return this.decoder;
  }

  public ICharacterEncoder GetEncoder() {
    return this.encoder;
  }
 }
